File: src/advent/2020/day6.clj
(ns advent.2020.day6
  "Advent of Code 2020, day 6: Custom Customs"
  (:require [advent.helpers :as h]
            [clojure.set :as set]))

(def puzzle-input (h/slurp-resource "2020/day6.txt" h/slurp-lines))

(defn group-items [items]
  (when (seq items)
    (let [[x xs] (split-with #(not= "" %) items)]
      (cons x (group-items (rest xs))))))

(defn puzzle1 [input]
  (->> (group-items input)
       (map (comp count #(reduce into #{} %)) )
       (apply +)))

(defn puzzle2 [input]
  (->> (group-items input)
       (map #(count (reduce set/intersection (map set %))))
       (apply +)))

File: src/haskell/2020/Day6.hs
{-|
Module      : Day6
Description : Day 6: Custom Customs

<https://adventofcode.com/2020/day/6>
-}

module Day6 where

import Advent
import Data.List (intersect)
import Data.Set (fromList)

main = do
    input <- parsedInput (2020, 6) (groupedLines . lines)
    print $ (sum . map (length . fromList . foldl1 (++))) input
    print $ (sum . map (length . foldl1 intersect)) input
